server/lib/scheduler/UserSchedule.ts

```diff
@@ -0,0 +1,75 @@
+import { Assignment } from "./Assignment";
+import { RootParentFinder } from "./RootParentFinder";
+
+class UserSchedule {
+  private timeLimit;
+  private userId;
+  private userSchedule = [];
+
+  public constructor({ timeLimit, userId }) {
+    this.timeLimit = timeLimit;
+    this.userId = userId;
+  }
+
+  public hasUserBeenAssignedToAnyWorkspaces() {
+    return this.userSchedule.length > 0;
+  }
+
+  public assignWorkspace(workspaceId, startAtTimestamp = Date.now()) {
+    const assignment = new Assignment({
+      userId: this.userId,
+      workspaceId,
+      startAtTimestamp,
+    });
+    this.userSchedule.push(assignment);
+  }
+
+  public getMostRecentAssignment() {
+    return this.userSchedule[this.userSchedule.length - 1];
+  }
+
+  public hasUserWorkedOnWorkspace(workspaceId) {
+    for (const assignment of this.userSchedule) {
+      if (assignment.getWorkspaceId() === workspaceId) {
+        return true;
+      }
+    }
+
+    return false;
```

Comment:
I think this can be cleaned up with lodash. Something like,
```
_.some(this.userSchedule, {workspaceId})
```
(I'm not sure if _.some takes a function, but you get the idea. That's also one disadvantage of hiding the workspaceId behind a method.)

Reply:
Actually, that read is very cheap, so if you had to first map, then _.some, that would probably be fine.

Reply:
Addressed in 99a270c